dtype for GDAL CInt16, rasterio complex_int16

scottyhq opened the pull request:
- [x] Closes #352
 - [x] Tests added
 - [x] Fully documented, including `docs/history.rst` for all changes and `docs/rioxarray.rst` for new API

Will follow up with a simple test and docs...

scottyhq commented:
@snowman2 would it be OK to put a small (100x100 Cint16) test here `test/test_data/input/cint16.tif` ? or is your preference for creating test datasets for example https://github.com/mapbox/rasterio/blob/master/tests/test_complex_dtypes.py#L37

snowman2 commented:
@scottyhq thanks for the fix 👍. Either way works for me. If you do add the raster, please compress it.

scottyhq commented on test/integration/test_integration__io.py:
```diff
@@ -1107,3 +1107,18 @@ def test_non_rectilinear__skip_parse_coordinates(open_rasterio):
+
+    # NOTE: waiting on https://github.com/mapbox/rasterio/issues/2206
+    # tmp_output = tmp_path / "tmp_cint16.tif"
+    # with pytest.warns(SerializationWarning):
+    #    xds.rio.to_raster(str(tmp_output), dtype='complex_int16')
+    #    with rasterio.open(str(tmp_output)) as riofh:
+    #        print(riofh.profile)
+    #        assert riofh.dtype == 'complex_int16'
```
Wanted to test the round trip to ensure CInt16 loaded with rioxarray is also written to a CInt16 tif and came across an issue with nodata handling https://github.com/mapbox/rasterio/issues/2206

snowman2 commented on docs/history.rst:
```diff
@@ -4,6 +4,10 @@ History
 
+0.4.3
+------
+- BUG: support GDAL CInt16, rasterio complex_int16 (pull #353)
```
This can go under the `Latest` section.

snowman2 commented on test/integration/test_integration__io.py:
```diff
@@ -1107,3 +1107,18 @@ def test_non_rectilinear__skip_parse_coordinates(open_rasterio):
+    # NOTE: waiting on https://github.com/mapbox/rasterio/issues/2206
+    # tmp_output = tmp_path / "tmp_cint16.tif"
+    # with pytest.warns(SerializationWarning):
+    #    xds.rio.to_raster(str(tmp_output), dtype='complex_int16')
```
I am thinking it would be good to introduce something like `_rasterio_dtype` in `encoding` and then check if the dtype is complex in `rio.to_raster` (just in case they change their dtype) and pass the value in the encoding to `rasterio` if it is.

scottyhq replied:
i attempted to do this, let me know if the revisions are what you had in mind

scottyhq commented on rioxarray/raster_writer.py:
```diff
@@ -155,7 +155,9 @@ def to_raster(self, xarray_dataarray, tags, windowed, lock, compute, **kwargs):
         **kwargs
             Keyword arguments to pass into writing the raster.
         """
-        dtype = kwargs["dtype"]
```
it isn't obvious to me where the default kwargs (dtype) comes from here (from a user's perspective calling this function without any additional arguments `da.rio.to_raster('raster.tif')`

snowman2 replied:
Here it is: https://github.com/corteva/rioxarray/blob/6204661cc9a703307207f548365b020a2d6eac59/rioxarray/raster_array.py#L916

snowman2 replied:
Unfortunately, with the complex dtype in rasterio, we will need to have a way to get the rasterio and numpy dtype. For the writer, I think that it is best to assume that dtype could be either the rasterio or numpy dtype and automatically determine the numpy/rasterio dtype from it.

scottyhq replied:
> Here it is: 

ah. thanks... still learning how all the pieces fit together!

scottyhq commented on rioxarray/raster_writer.py:
```diff
@@ -155,7 +155,9 @@ def to_raster(self, xarray_dataarray, tags, windowed, lock, compute, **kwargs):
             Keyword arguments to pass into writing the raster.
         """
-        dtype = kwargs["dtype"]
+        numpy_dtype = kwargs["dtype"]
+        if xarray_dataarray.encoding.get("_rasterio_dtype") == "complex_int16":
+            kwargs["dtype"] = "complex_int16"
```
> we will need to have a way to get the rasterio and numpy dtype. For the writer, I think that it is best to assume that the "dtype" kwarg could be either the rasterio or numpy dtype and automatically determine the numpy/rasterio dtype from it. 

i think the only special dtype is `complex_int16` so i went with this approach, let me know if you prefer something else. tests pass except for the new one (requires ` pip install git+https://github.com/mapbox/rasterio.git@maint-1.2`, fixes aren't yet in the master branch

snowman2 replied:
I think the fixes are in master now.

snowman2 commented on rioxarray/_io.py:
```diff
@@ -882,6 +886,10 @@ def open_rasterio(
+    encoding["_rasterio_dtype"] = str(riods.dtypes[0])
+    # .encoding["dtype"] used by xarray write methods
     if masked:
         encoding["dtype"] = str(riods.dtypes[0])
```
I think the `encoding["dtype"]` needs the check from above here as well:
```python
        if riods.dtypes[0] == "complex_int16":
                encoding["dtype"] = np.dtype("complex64")
...
```
Might be worthwhile to add a helper function to get the numpy dtype from rasterio dtype.

snowman2 commented on rioxarray/raster_writer.py:
```diff
@@ -155,7 +155,9 @@ def to_raster(self, xarray_dataarray, tags, windowed, lock, compute, **kwargs):
         """
-        dtype = kwargs["dtype"]
+        numpy_dtype = kwargs["dtype"]
+        if xarray_dataarray.encoding.get("_rasterio_dtype") == "complex_int16":
```
I am thinking we should check if the numpy dtype is complex in addition to the rasterio dtype check here. Just in case they did an `astype(...)` or passed in a different dtype to convert it to something else.

snowman2 commented:
@scottyhq thanks for you patience with this. It is looking good. I added a couple of comments for things that would be good to consider.

scottyhq commented:
@snowman2 I added more tests to cover dealing with masking and nodata. Given the single `dtype=` kwarg in to_raster which by default matches the in-memory np.dtype, and the fact that CInt16 is a special case, I think it's easiest to just require an explicit dtype='complex_int16' if you want to *write* to that format. So I added a note in the docstring. 

Tests pass for me locally, except for `test_integration_merge.py`, which seems unrelated to this PR, and is due to rounding changes in rasterio>1.2.5: (https://github.com/mapbox/rasterio/issues/2202)

snowman2 commented:
@scottyhq this looks great, thanks :+1:

snowman2 commented:
@scottyhq any issues if I squash and merge?

scottyhq commented:
> @scottyhq any issues if I squash and merge?

Fine by me! Thanks for the guidance @snowman2
